README.md
## movee-log
> 利用阿里云web-track记录前端日志，并支持微信小程序和web应用

![npm](https://img.shields.io/npm/dt/movee-log) ![npm](https://img.shields.io/npm/v/movee-log) ![npm](https://img.shields.io/npm/l/movee-log) ![GitHub stars](https://img.shields.io/github/stars/HongYangHT/movee-log?style=social) ![GitHub forks](https://img.shields.io/github/forks/HongYangHT/movee-log?style=social)

### 使用方法
> 查看[使用方法](./USEAGE.md)

### PR && Commit (提交代码)
- 使用 commitizen 来格式化 Git commit message
  - 安装 commitizen
  > npm install -g commitizen
  - 使用 angular 的 commit 规范 commitizen init cz-conventional-changelog --save-dev --save-exact
  - 重新安装 husky 与 lint-staged
  > npm i -D husky  
  > npm i -D lint-staged
  - 使用 `git cz` 代替 commit

- 提交规则
  - feat: 新功能
  - fix: 修复bug
  - docs: 文档更新
  - style: 格式更新（不影响代码运行的变动）
  - refactor: 重构（既不是新增功能，又不是bug修复）
  - test: 添加测试
  - chore: 构建过程或辅助工具的变动

- git 提交
  - 将 `git commit` 用 `git cz` 代替 

### 更新日志
> 查看更新[日志](./CHANGELOG.md)

- 生成更新日志
> npm install -g conventional-changelog-cli

- 发布新版本推荐工作流
  1. 执行eslint格式化代码 (npm run eslint)
  2. 提交修改的commit (git add . / git cz)
  3. 修改版本号 (npm run changeVersion)
  4. 执行build (npm run build)
  5. 自动生成change log (npm run changelog)
  6. 重新添加package.json 和 CHANGELOG.md (git add . / git cz)
  7. 打tag (git add -a tag -m 'xx')
  8. push (git push origin tag)
  9. 发布到npm


CHANGELOG.md
## [1.0.10](https://github.com/HongYangHT/movee-log/compare/1.0.1...1.0.10) (2019-12-28)



## [1.0.9](https://github.com/HongYangHT/movee-log/compare/1.0.0...1.0.9) (2019-12-19)


### Features

* 修改版本号 ([3b2b7ae](https://github.com/HongYangHT/movee-log/commit/3b2b7aefddfbb482a81189e7da141e9331c704a2))



## [1.0.8](https://github.com/HongYangHT/movee-log/compare/1.0.0...1.0.8) (2019-12-17)



## [1.0.7](https://github.com/HongYangHT/movee-log/compare/1.0.2...1.0.7) (2019-12-17)


### Bug Fixes

* 抛出Logger对象 ([ce538a6](https://github.com/HongYangHT/movee-log/commit/ce538a6803e26a2438da031da7091a2a65875362))


### Features

* 发布新版本 ([e207652](https://github.com/HongYangHT/movee-log/commit/e20765202582bd8efb9d71d114f264fdb7be2e36))
* 发布新版本 ([641ae54](https://github.com/HongYangHT/movee-log/commit/641ae54f8b6b216983e956295210f09303863316))



## [1.0.6](https://github.com/HongYangHT/movee-log/compare/1.0.2...1.0.6) (2019-12-17)


### Bug Fixes

* 抛出Logger对象 ([ce538a6](https://github.com/HongYangHT/movee-log/commit/ce538a6803e26a2438da031da7091a2a65875362))


### Features

* 发布新版本 ([641ae54](https://github.com/HongYangHT/movee-log/commit/641ae54f8b6b216983e956295210f09303863316))



## [1.0.5](https://github.com/HongYangHT/movee-log/compare/1.0.2...1.0.5) (2019-12-17)


### Bug Fixes

* 抛出Logger对象 ([ce538a6](https://github.com/HongYangHT/movee-log/commit/ce538a6803e26a2438da031da7091a2a65875362))



## [1.0.4](https://github.com/HongYangHT/movee-log/compare/1.0.2...1.0.4) (2019-12-17)


### Bug Fixes

* 抛出Logger对象 ([ce538a6](https://github.com/HongYangHT/movee-log/commit/ce538a6803e26a2438da031da7091a2a65875362))



## [1.0.3](https://github.com/HongYangHT/movee-log/compare/1.0.2...1.0.3) (2019-12-17)


### Bug Fixes

* 抛出Logger对象 ([ce538a6](https://github.com/HongYangHT/movee-log/commit/ce538a6803e26a2438da031da7091a2a65875362))



## 1.0.2 (2019-12-17)


### Bug Fixes

* 修改源码 ([fc31694](https://github.com/HongYangHT/movee-log/commit/fc31694fd2e487a377dce3fb11dd75e6a316debe))


### Features

* init project ([340829e](https://github.com/HongYangHT/movee-log/commit/340829e014692de9befae4b5e23fcfdff0eaf35c))
* 修改打包和eslit配置 ([c0ba4eb](https://github.com/HongYangHT/movee-log/commit/c0ba4eb2eed3e8422094503843c003f4c111b66f))
* 修改相关配置 ([159af60](https://github.com/HongYangHT/movee-log/commit/159af60db0dfd82bca0456ea4fa034f745d76e56))
* 合并代码 ([baf45c9](https://github.com/HongYangHT/movee-log/commit/baf45c960a35bc3db00b92d884c563fa167a6856))
* 新增使用的md方法 ([3a1a6e0](https://github.com/HongYangHT/movee-log/commit/3a1a6e0b2171c0402597fc27184d98d79de5af1d))



## 1.0.1 (2019-12-17)


### Bug Fixes

* 修改源码 ([fc31694](https://github.com/HongYangHT/movee-log/commit/fc31694fd2e487a377dce3fb11dd75e6a316debe))


### Features

* init project ([340829e](https://github.com/HongYangHT/movee-log/commit/340829e014692de9befae4b5e23fcfdff0eaf35c))
* 修改打包和eslit配置 ([c0ba4eb](https://github.com/HongYangHT/movee-log/commit/c0ba4eb2eed3e8422094503843c003f4c111b66f))
* 修改相关配置 ([159af60](https://github.com/HongYangHT/movee-log/commit/159af60db0dfd82bca0456ea4fa034f745d76e56))
* 合并代码 ([baf45c9](https://github.com/HongYangHT/movee-log/commit/baf45c960a35bc3db00b92d884c563fa167a6856))





.eslintrc.js
/*
 * @Author: sam.hongyang
 * @LastEditors: sam.hongyang
 * @Description: eslint 设置
 * @Date: 2019-12-17 14:37:38
 * @LastEditTime: 2019-12-17 17:47:15
 */
module.exports = {
  root: true,
  parserOptions: {
    parser: 'babel-eslint',
    "sourceType": "module",
    "allowImportExportEverywhere": true
  },
  env: {
    browser: true,
    es6: true
  },
  "extends": ["prettier"],
  "plugins": ["prettier"],
  "rules": {
    "prettier/prettier": ["error",
    {
      printWidth: 100,
      singleQuote: true,
      trailingComma: 'none',
      bracketSpacing: true,
      semi: false,
      tabWidth: 2,
      useTabs: false,
      alwaysParens: 'avoid'
    }],
    // allow async-await
    'generator-star-spacing': 'off',
    // allow debugger during development
    'no-debugger': process.env.NODE_ENV === 'production' ? 'error' : 'off',
    'no-console': process.env.NODE_ENV === 'production' ? 'error' : 'off',
    camelcase: "off", // 强制驼峰法命名
    "no-new": "off", // 禁止在使用new构造一个实例后不赋值
    "space-before-function-paren": "off", // 函数定义时括号前面不要有空格
    "no-plusplus": "off", // 禁止使用 ++， ——
    "max-len": "off", // 字符串最大长度
    "func-names": "off", // 函数表达式必须有名字
    "no-param-reassign": "off" // 不准给函数入参赋值
  }
}


src/track.js
/*
 * @Author: sam.hongyang
 * @LastEditors: sam.hongyang
 * @Description: track.js, 用于发送日志请求
 * @Date: 2019-12-11 17:02:00
 * @LastEditTime: 2019-12-11 17:05:24
 */
(function (window, document) {
    function createHttpRequest() {
        if (window.ActiveXObject) {
            return new ActiveXObject("Microsoft.XMLHTTP");
        } else if (window.XMLHttpRequest) {
            return new XMLHttpRequest();
        }
    }

    function AliLogTracker(host, project, logstore) {
        this.uri_ = 'http://' + project + '.' + host + '/logstores/' + logstore + '/track?APIVersion=0.6.0';
        this.params_ = new Array();
        this.httpRequest_ = createHttpRequest();
    }
    AliLogTracker.prototype = {
        push: function (key, value) {
            if (!key || !value) {
                return;
            }
            this.params_.push(key);
            this.params_.push(value);
        },
        logger: function () {
            var url = this.uri_;
            var k = 0;
            while (this.params_.length > 0) {
                if (k % 2 == 0) {
                    url += '&' + encodeURIComponent(this.params_.shift());
                } else {
                    url += '=' + encodeURIComponent(this.params_.shift());
                }
                ++k;
            }
            try {
                this.httpRequest_.open("GET", url, true);
                this.httpRequest_.send(null);
            } catch (ex) {
                if (window && window.console && typeof window.console.log === 'function') {
                    console.log("Failed to log to ali log service because of this exception:\n" + ex);
                    console.log("Failed log data:", url);
                }
            }
        }
    };
    window.Tracker = AliLogTracker;
})(window, document);

webpack.config.js
/*
 * @Author: sam.hongyang
 * @LastEditors: sam.hongyang
 * @Description: webpack 相关配置
 * @Date: 2019-12-11 17:14:42
 * @LastEditTime: 2019-12-17 18:06:01
 */
const pkg = require('./package.json')
const webpack = require('webpack')
const path = require('path')
const { CleanWebpackPlugin } = require('clean-webpack-plugin')
const SpeedMeasurePlugin = require('speed-measure-webpack-plugin')
const CompressionWebpackPlugin = require('compression-webpack-plugin')

const banner = `
${pkg.name}
${pkg.description}\n
@version v${pkg.version}
@homepage ${pkg.homepage}
@repository ${pkg.repository.url}\n
(c) 2019 sam.hongyang
Released under the MIT License.
hash: [hash]
`
const plugins = [
  new webpack.BannerPlugin(banner),
  new CleanWebpackPlugin({
    cleanOnceBeforeBuildPatterns: ['**/*'],
    verbose: true,
    dry: false
  }),
  new CompressionWebpackPlugin({
    filename: '[path].gz[query]',
    algorithm: 'gzip',
    test: /\.js$|\.css$|\.html$|\.eot?.+$|\.ttf?.+$|\.woff?.+$|\.svg?.+$/,
    // NOTE: 设置生成`gzip`文件大小 100K
    threshold: 100 * 1024,
    minRatio: 0.8,
    cache: true
  })
]

const smp = new SpeedMeasurePlugin()

module.exports = smp.wrap({
  mode: 'production',
  entry: {
    index: './index.js'
  },
  output: {
    path: path.resolve(__dirname, 'dist'),
    libraryTarget: 'umd',
    filename: '[name].js',
    chunkFilename: '[name].js',
    umdNamedDefine: true
  },
  // context是webpack编译时的基础目录，entry会相对此目录查找
  // path.join连接路径，将几个路径拼接到一起
  // process.cwd()是当前执行node命令的地址  __dirname是当前模块的目录
  context: path.join(process.cwd(), 'src'),
  resolve: {
    modules: ['node_modules', path.resolve(process.cwd(), 'src')],
    alias: {},
    extensions: ['.js']
  },
  module: {
    rules: [
      {
        test: /\.(js|vue)$/,
        loader: 'eslint-loader',
        enforce: 'pre',
        include: [path.join(__dirname, 'src')],
        options: {
          fix: true
        }
      },
      {
        test: /\.js$/,
        use: ['babel-loader'],
        exclude: /node_modules/
      }
    ]
  },
  performance: {
    hints: false
  },
  // optimization: {
  //   // 根据不同的策略来分割打包出来的bundle
  //   splitChunks: {
  //     // 同时分割同步和异步代码
  //     chunks: 'all'
  //   },
  //   // 对于每个entry生成对应的runtime~${entrypoint.name}文件
  //   runtimeChunk: true
  // },
  plugins
})


src/index.js
/*
 * @Author: sam.hongyang
 * @LastEditors  : sam.hongyang
 * @Description: 发送请求日志
 * @Date: 2019-12-16 16:49:15
 * @LastEditTime : 2019-12-28 14:29:17
 */
class Logger {
  /**
   * 配置相关参数
   * @param {string} project 项目前缀
   * @param {string} host 项目的host
   * @param {string} logstore 项目日志存存的位置
   * @param {string} from 来源的项目
   */
  constructor(
    project,
    host,
    logstore,
    accessId,
    accessKey,
    from,
    ua = 'web',
    projectEnv = 'dev',
    openLog = true
  ) {
    this.project = project
    this.host = host
    this.logstore = logstore
    this.from = from
    this.ua = ua // NOTE: uni:表示builx 编译的uni-app小程序 wx: 表示直接用微信小程序编写的小程序 web: 表示web应用（包含web和H5）
    this.accessId = accessId // NOTE: access id
    this.accessKey = accessKey // NOTE: access key
    this.uri = `https://${project}.${host}/logstores/${logstore}/track?APIVersion=0.6.0` // NOTE: 发送给阿里云的后台接口地址
    this.projectEnv = projectEnv
    this.openLog = openLog
  }

  padLeftZero(str) {
    return ('00' + str).substr(str.length)
  }

  formatDate(d, format = 'yyyy-MM-dd hh:mm:ss') {
    let date = new Date(d)
    if (/(y+)/.test(format)) {
      format = format.replace(RegExp.$1, (date.getFullYear() + '').substr(4 - RegExp.$1.length))
    }
    let o = {
      'M+': date.getMonth() + 1,
      'd+': date.getDate(),
      'h+': date.getHours(),
      'm+': date.getMinutes(),
      's+': date.getSeconds()
    }
    for (let k in o) {
      if (new RegExp(`(${k})`).test(format)) {
        let str = o[k] + ''
        format = format.replace(RegExp.$1, RegExp.$1.length === 1 ? str : this.padLeftZero(str))
      }
    }
    return format
  }

  /**
   * 查询当前环境的ua
   * 1. 操作系统
   * 2. 应用的版本号，如浏览器版、小程序本号等
   */
  fetchNavigation() {
    let info = {}
    switch (this.ua) {
      case 'uni':
        try {
          uni.getSystemInfo({
            success: res => {
              info = Object.assign(info, {
                os: res.platform,
                browser: `${res.brand}/${res.model}/${res.system}/${res.SDKVersion}/${res.version}`
              })
            }
          })
        } catch (error) {
          console.log(error)
        }
        break
      case 'wx':
        try {
          wx.getSystemInfo({
            success: res => {
              info = Object.assign(info, {
                os: res.platform,
                browser: `${res.brand}/${res.model}/${res.system}/${res.SDKVersion}/${res.version}`
              })
            }
          })
        } catch (error) {
          console.log(error)
        }
        break
      case 'web':
        try {
          let navi = window.navigator
          info = Object.assign(info, {
            os: navi.platform,
            browser: navi.userAgent
          })
        } catch (error) {
          console.log(error)
        }
        break
      default:
        try {
          let navi = window.navigator
          info = Object.assign(info, {
            os: navi.platform,
            browser: navi.userAgent
          })
        } catch (error) {
          console.log(error)
        }
        break
    }
    return info
  }

  /**
   * 获取用户信息
   */
  fetchUserInfo() {
    let info = {}
    switch (this.ua) {
      case 'uni':
        info = Object.assign(info, {
          token: uni.getStorageSync('token') || '',
          userId: uni.getStorageSync('userId') || '',
          username: uni.getStorageSync('loginName') || ''
        })
        break
      case 'wx':
        info = Object.assign(info, {
          token: wx.getStorageSync('token') || '',
          userId: wx.getStorageSync('userId') || '',
          username: wx.getStorageSync('loginName') || ''
        })
        break
      case 'web':
        info = Object.assign(info, {
          token: window.localStorage.getItem('token') || '',
          userId: window.localStorage.getItem('userId') || '',
          username: window.localStorage.getItem('loginName') || ''
        })
        break
      default:
        info = Object.assign(info, {
          token: window.localStorage.getItem('token') || '',
          userId: window.localStorage.getItem('userId') || '',
          username: window.localStorage.getItem('loginName') || ''
        })
        break
    }
    return info
  }
  /**
   * 发送日志请求
   * @param {string} logType 日志类型 request: 请求日志 request-error: 请求失败的日志
   * @param {string} requestUrl 请求的地址
   * @param {string} params 请求的入参
   * @param {string} requestTime 请求执行的时间
   * @param {string} url 发送日志的路由页面
   * @param {string} responseData 请求失败后返回的参数
   * @param {string} stack 错误日志堆栈
   */
  logger(
    logType = 'request',
    requestUrl,
    requestType,
    params,
    requestTime,
    url,
    responseData,
    stack
  ) {
    let userInfo = this.fetchUserInfo()
    let uaInfo = this.fetchNavigation()
    let urlUri = this.uri
    let info = Object.assign(userInfo, uaInfo, {
      logType,
      happenTime: this.formatDate(new Date(), 'yyyy-MM-dd hh:mm:ss'), // NOTE: 接口发送的时间
      params: JSON.stringify(params),
      url,
      requestUrl,
      requestType,
      requestTime,
      responseData: JSON.stringify(responseData),
      projectEnv: this.projectEnv,
      stack: JSON.stringify(stack)
    })
    this.request(urlUri, info)
  }

  jsLogger(logType, url, stack, sourceUrl) {
    let userInfo = this.fetchUserInfo()
    let uaInfo = this.fetchNavigation()
    let urlUri = this.uri
    let info = Object.assign(userInfo, uaInfo, {
      logType,
      happenTime: this.formatDate(new Date(), 'yyyy-MM-dd hh:mm:ss'),
      url,
      projectEnv: this.projectEnv,
      stack: JSON.stringify(stack),
      sourceUrl
    })
    this.request(urlUri, info)
  }

  /**
   * 发送请求的函数
   * @param {string} url 请求的url
   * @param {object} params 请求的参数
   */
  request(url, params) {
    if (!this.openLog) return
    let urlUri = url
    let keys = Object.keys(params)
    let xhr = null
    keys.forEach(key => {
      if (key && params[key]) {
        urlUri += `&${key}=${encodeURIComponent(params[key])}`
      }
    })

    switch (this.ua) {
      case 'uni':
        try {
          uni.request({
            method: 'GET',
            url: urlUri,
            header: {
              accessId: this.accessId,
              accessKey: this.accessKey
            }
          })
        } catch (error) {
          console.log(error)
          console.log(error)
        }
        break
      case 'wx':
        try {
          wx.request({
            method: 'GET',
            url: urlUri,
            header: {
              accessId: this.accessId,
              accessKey: this.accessKey
            }
          })
        } catch (error) {
          console.log(error)
        }
        break
      case 'web':
        if (window.ActiveXObject) {
          // eslint-disable-next-line
          xhr = new ActiveXObject('Microsoft.XMLHTTP')
        } else if (window.XMLHttpRequest) {
          xhr = new XMLHttpRequest()
        }

        try {
          xhr.open('GET', urlUri, true)
          xhr.setRequestHeader('accessId', this.accessId)
          xhr.setRequestHeader('accessKey', this.accessKey)
          xhr.send(null)
        } catch (ex) {
          if (window && window.console && typeof window.console.log === 'function') {
            console.log('Failed to log to log service because of this exception:\n' + ex)
            console.log('Failed log data:', url)
          }
        }
        break
      default:
        if (window.ActiveXObject) {
          // eslint-disable-next-line
          xhr = new ActiveXObject('Microsoft.XMLHTTP')
        } else if (window.XMLHttpRequest) {
          xhr = new XMLHttpRequest()
        }

        try {
          xhr.open('GET', urlUri, true)
          xhr.setRequestHeader('accessId', this.accessId)
          xhr.setRequestHeader('accessKey', this.accessKey)
          xhr.send(null)
        } catch (ex) {
          if (window && window.console && typeof window.console.log === 'function') {
            console.log('Failed to log to log service because of this exception:\n' + ex)
            console.log('Failed log data:', url)
          }
        }
        break
    }
  }
}

export default Logger
